docs/embroider-architecture-overview.md

```diff
@@ -0,0 +1,402 @@
+
+So, the ultimate goal of Embroider is to be able to point a modern JS build tool at your real Ember app folder, instead of a version of it that has been copied to a tmp directory.
+
+## The App
+
+As it stands, a very basic Ember app might look like this (for simplicity, let's ignore things likes `/tests` and `/styles`):
+
+```bash
+/
+├── app
+│   ├── index.html
+│   ├── app.js
+│   ├── components
+│   │   ├── user-card.js
+│   │   └── user-card.hbs
+│   └── templates
+│       └── application.hbs
+├── ember-cli-build.js
+└── package.json
+```
+
+If we look at the `index.html` file, we can quickly see why pointing a modern build tool at it won't work.
+
+```html
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8" />
+    <link rel="stylesheet" href="/assets/vendor.css" />
+    <link rel="stylesheet" href="/assets/my-app.css" />
+  </head>
+  <body>
+    <script src="/assets/vendor.js"></script>
+    <script src="/assets/my-app.js"></script>
+  </body>
+</html>
+```
+
+Unlike the "perfect world" example above, this `index.html` points to `.js` and `.css` files that don't actually exist. This is because the are pointing to files in a location that will be created as a part of the EmberCLI build logic that runs when we run the `ember build` command. Let's look a bit deeper to see what else we can find.
+
+## The Addons
+**TODO**
+
+## Embroider Build Stages
+
+### Stage 1 - Compat Addons
+
+This stage is all about ensuring our Ember addon dependencies conform to the V2 Package Format ([RFC](https://rfcs.emberjs.com/id/0507-embroider-v2-package-format/)). This stage goes through each of the EmberApp's dependencies, and their dependencies to check if they need to be re-written to the V2 Package format. Because the addons can't be modified in place, in the project's `node_modules` folder, they are written to a tmp Embroider directory. So, Embroider effectively recreates the `node_modules` directory in this tmp dir. For dependencies that do need to be converted to V2, it creates a new copy of the dependency. For those that do not need any modification, they are symlinked in to the tmp dir.
+
+The following checks determine if a dependency does not need to be rewritten:
+
+- The dependency is not an Ember addon
+- The dependency is a V2 Ember addon _and_ none of its own dependencies need to be rewritten
+
+The following checks determine if a dependency does need to be rewritten:
+
+- The dependency is a V1 Ember addon or;
+- The dependency is a V2 Ember addon but one or more of it's dependencies must be rewritten (ie, are a V1 addon)
+
+Or, put another way, as a picture:
+
+```mermaid
+flowchart TD
+    A[Dependency] --> B{Is Ember addon?}
+    B -->|Yes| C{Is V2 addon?}
+    C -->|Yes| D{Has deps\nthat need\nrewriting?}
+    D -->|No| F[Symlink]
+    D -->|Yes| E[Rewrite] 
+    C -->|No| E[Rewrite]
+    B ---->|No| F[Symlink]
```

Comment:
By reordering lines 134-140, we can avoid overlapping edges, which may confuse the reader.

If we pretend that symlink is good and rewrite is bad, I think we could add the ✅ and ❌ icons to hint which path the reader will want to take in the flowchart.

| Before | After |
|:--:|:--:|
| <img width="300" alt="" src="https://github.com/embroider-build/embroider/assets/16869656/9286b70f-a61d-4cd7-a539-9f8452b5841a"> | <img width="300" alt="" src="https://github.com/embroider-build/embroider/assets/16869656/bed626cb-8caa-4406-8a6b-8df05ef78c07"> |

```suggestion
flowchart TD
    A[Dependency] --> B{Is Ember addon?}
    B -->|Yes| C{Is V2 addon?}
    C -->|No| E["❌ Rewrite"]
    C -->|Yes| D{Has dependency\nthat needs rewrite?}
    D -->|Yes| E["❌ Rewrite"]
    D -->|No| F["✅ Symlink"]
    B -->|No| F["✅ Symlink"]
```

Reply:
Great suggestion. Thanks @ijlee2 .